proposals/NNNN-fix-expressible-by-string-interpolation.md

```diff
@@ -0,0 +1,822 @@
+//             ^~~~~~
+//    String<T>(_: T) where T: LosslessStringConvertible
+  
+print("Hello, \(name as Any)")
+//             ^~~~~~~~~~~~~
+//    String<T>(stringInterpolationSegment: T)
+```
+	
+  [se0089]: https://github.com/apple/swift-evolution/blob/master/proposals/0089-rename-string-reflection-init.md
+
+#### Rationale for use of initializers
+
+The `StringInterpolationType` acts as a sort of filter or funnel which 
+controls which interpolations are permitted and marshals them into a 
+single type that `init(stringInterpolation:)` 
+can handle in a uniform way. At the same time, because 
+`StringInterpolationType` is not constrained by a formal protocol, 
+the type has infinite flexibility in *what* it permits to interpolate 
+and *how* it handles that input. It can overload to support different 
+types, it can use argument labels to request special handling, and it 
+can take additional parameters to customize the result. It is also free 
+to *reject* certain inputs by not providing an initializer that accepts 
+them; the Swift compiler will then diagnose those errors at compile time.
+
+By having this be a separate associated type, rather than using 
+initializers on `Self`, types can delegate the job to another type. In 
+particular, they can delegate it to `String`, which has a wide variety 
+of useful initializers and is a common extension point for additional 
+ones. We expect that many types will use `String`, especially if they 
+handle arbitrary text. But those types which need precise control over 
+interpolation can designate themselves or a helper type instead.
+
+We do not propose that every formatting feature Swift supports should 
+be accessed by adding labeled parameters to initializers. But 
+we believe that feeding interpolated values through an initializer on 
+an associated type is a good primitive for formatting features. For 
+instance, [here's a sketch of a formatting DSL][formatsketch] 
+compatible with this proposal which lets you say things like:
+
+```swift
+"You are number \(ticketNumber %% int(sign: .never) %% width(padding: .leading, to: 5))"
+```
+
+  [formatsketch]: https://gist.github.com/brentdax/c3f9dc7c03a34c979e5e363826c2e19d
+
+We show this example not to propose this design, but rather to 
+demonstrate that what we *are* proposing here has the flexibility to 
+accommodate many formatting designs. We leave the precise design of 
+such a feature to future proposals and community-led projects.
+
+We also believe that, even without any additional formatting features, 
+initializers are an adequate solution for the code-generation use 
+case, where you don't want the flexibility of a general-purpose 
+formatting system:
+
+```swift
+extension SQLStatement {
+  init(raw sql: String) { ... }
+  init(raw sql: SQLStatement) { ... }
+  init(_ parameter: SQLValue) { ... }
+}
+
+extension SQLStatement: ExpressibleByStringLiteral {
+  typealias StringInterpolationType = SQLStatement
```

Comment:
Shouldn't that protocol conformance be `ExpressibleByStringInterpolation` here, not `ExpressibleByStringLiteral`?
Especially since the `typealias StringInterpolationType` is supposed to be defining the `associatedtype` on `ExpressibleByStringInterpolation`, not on `ExpressibleByStringLiteral`